auth/src/main/java/org/aerogear/android/ags/auth/utils/AuthStateManager.java

```diff
@@ -1,60 +1,75 @@
-package org.aerogear.auth.utils;
+package org.aerogear.android.ags.auth.utils;
 
 import android.content.Context;
 import android.content.SharedPreferences;
-import org.aerogear.auth.credentials.OIDCCredentials;
-import org.json.JSONException;
+
+import org.aerogear.android.ags.auth.credentials.OIDCCredentials;
 
 /**
  * Saves, retrieves and delete a token.
  */
 public class AuthStateManager {
 
+    private static AuthStateManager instance = null;
     private static final String STORE_NAME = "org.aerogear.android.auth.AuthState";
     private static final String KEY_STATE = "state";
 
     private final SharedPreferences prefs;
 
-    public AuthStateManager(final Context context) {
+    private AuthStateManager(final Context context) {
         this.prefs = context.getSharedPreferences(STORE_NAME, Context.MODE_PRIVATE);
     }
 
     /**
      * Reads credentials from storage.
      * @return OIDCCredentials
      */
-    public OIDCCredentials read() {
-        String currentState = prefs.getString(KEY_STATE, null);
+    public OIDCCredentials load() {
+        final String currentState = prefs.getString(KEY_STATE, null);
         if (currentState == null) {
             return new OIDCCredentials();
         }
-        try {
-            return new OIDCCredentials(currentState);
-        } catch (JSONException ex) {
-            return new OIDCCredentials();
-        }
+        return OIDCCredentials.deserialize(currentState);
     }
 
     /**
      * Saves a token
      * @param authState token to be saved
+     * @throws IllegalStateException
      */
-    public synchronized void write(final OIDCCredentials authState) {
+    public synchronized void save(final OIDCCredentials authState) {
         if (authState == null) {
             clear();
         } else {
-            if(!prefs.edit().putString(KEY_STATE, authState.serialise()).commit()) {
+            SharedPreferences.Editor e = prefs.edit();
+            SharedPreferences.Editor bleh = e.putString(KEY_STATE, authState.serialize());
+            if(!bleh.commit()) {
                 throw new IllegalStateException("Failed to update state from shared preferences");
             }
         }
     }
 
     /**
      * Deletes a token
+     * @throws IllegalArgumentException
      */
     public synchronized void clear() {
         if (!prefs.edit().remove(KEY_STATE).commit()) {
             throw new IllegalStateException("Failed to clear state from shared preferences");
         }
     }
+
+    public static AuthStateManager getInstance(final Context context) {
```

Comment:
@rachael-oregan To respond to this comment https://github.com/aerogear/aerogear-android-sdk/pull/21/commits/260a5b358f4678d653b940a4d4fab4738b4d6bcc#r166298337 yeah I'm not sure the best way to approach this. I don't think providing a 'setter' of any kind is a good idea as then context could change if the setter is called multiple times. This can't be initialized multiple times but I can see the awkwardness of it.

This way context is bound to whatever is provided first, instead of being able to be set multiple times.

There's also the option of providing context on each function call I guess.

Perhaps @ziccardi or @secondsun have a more informed view on how to approach this.

Reply:
@aidenkeating Just a proposal: we could add a `bootstrap` method with package level visibility. It could then be called from the `bootstrap` method of the AuthService. This way other classes will see only a `getInstance` with no parameters. WDYT?